### ISSUE TITLE
Destruction of TypeScript ecosystem on npm is expanding

### ISSUE THREAD

**Issue by falsandtru** [microsoft/TypeScript#32273]
As I mentioned at #31148, unused @types files installed by node modules pollutes and breaks the main project. It is caused by del package at that time. *Now this problem came to be spread by another major package npm-check-updates.* **This means the destruction of ecosystem is expanding.** TypeScript must resolve the problem quickly. See a solution https://github.com/microsoft/TypeScript/issues/18588#issuecomment-492616933.

**Comment by falsandtru** [comment]
My proposal is https://github.com/microsoft/TypeScript/issues/18588#issuecomment-492616933. You just implement it.

**Comment by ExE-Boss** [comment]
I&nbsp;believe that the&nbsp;best option is&nbsp;to&nbsp;just use&nbsp;[pnpm](https://pnpm.js.org), which&nbsp;doesn’t have this problem, as&nbsp;packages are&nbsp;stored in&nbsp;isolated trees and&nbsp;are&nbsp;symlinked to&nbsp;each other.

---

Typical **pnpm** file hierarchy:
```
node_modules/
- .registry.npm.com/
	- @types/
		- glob/
		- node/
	- del/
		- 4.1.1/
			- node_modules/
				- del/
				- @types/
					- glob - ../../../../@types/glob/<version>/node_modules/@types/glob
- del - .registry.npm.com/del/4.1.1/node_modules/del
```

**Comment by StreetStrider** [comment]
@ExE-Boss, hello. I came to this ticket by the opposite reason. I'm already using pnpm actively in my non-TS projects. However, when I try to install `@types` package in pnpm (with classic deps tree hierarchy) tsc rejects to resolve transient dependencies of such `@types` packages. Just to check my guess, I tried reinstalling with vanilla npm (flat deps). From that point tsc works OK, which means no errors in my typed code. However when I switch back with `pnpm install` tsc starts to nag again, like:
```
node_modules/@types/express/index.d.ts:19:23 - error TS2307: Cannot find module 'express-serve-static-core'.

19 import * as core from "express-serve-static-core";
```
(I depend on `@types/express`).
Have you faced similar issues (in past or present)? Is it my own issue which I should investigate further or tsc now hard depends on flat order of dependecies?
